### ISSUE TITLE
Undefined index: permissions at .../groupfolders/lib/Versions/VersionsBackend.php (line #155)

### ISSUE THREAD

**Issue by reinob** [nextcloud/groupfolders#266]
In Nextcloud log file the following is logged:
```
{"reqId":"XTfYF2roLTToVh6q49pu","level":3,"time":"2018-12-17T06:38:02+00:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":"Undefined index: permissions at \/var\/www\/nextcloud\/apps\/groupfolders\/lib\/Versions\/VersionsBackend.php#155","userAgent":"--","version":"15.0.0.10"}
```
It is (yet) unclear what triggers the error. Besides cron.php I periodically run "files:scan --all" and "preview:pre-generate" but none of the three cronjobs appears to trigger that error every time, yet, it happens (always 3 times) every hour or so. The 3 times comes probably from the fact that I have 3 group folders defined.

Server configuration
Debian stretch, Apache, PostgreSQL 11, PHP 7.2, Nextcloud 15 (fresh install)

The offending line is:
```
$mount = $this->mountProvider->getMount($folder['id'], '/dummyuser/files/' . $folder['mount_point'], $folder['permissions'], $folder['quota']);
```
here it should be checked whether $folder['permissions'] exists at all before trying to use it.

**Comment by reinob** [comment]
I still don't know the trigger for this error, but it happens when the cronjon cron.php is run, but not every time.

I can also trigger a variant of it with occ groupfolders:can

```
{"reqId":"dgWYd8PBLy08KKzY0LxX","level":3,"time":"2018-12-22T13:10:01+00:00","remoteAddr":"","user":"--","app":"PHP","method":"","url":"--","message":"Undefined index: permissions at \/var\/www\/nextcloud\/apps\/groupfolders\/lib\/Command\/Scan.php#59","userAgent":"--","version":"15.0.0.10"}
```

I will try adding array_key_exists here and there and see if that does it.

**Comment by reinob** [comment]
OK, in lib/Versions/VersionsBackend.php, line 155 I replaced `$folder['permissions']` with `array_key_exists('permissions', $folder) ? $folder['permissions'] : null`

in lib/Command/Scan.php, line 59 I did the same.

For now the logs seem to be quiet..

(PHP really sucks, but using variables without being sure that they exist sucks too..)

**Comment by ntinti** [comment]
Bump...
Same problem here

**Comment by landryb** [comment]
Also seeing it here with 2.0.3 - @icewind1991 do you think checking if the key exists in $folder enough of a fix, or that's hiding a real issue ? can do a PR if needed, but that smells fishy.

The log seems trigerred at regular intervals (quarter of hour most of the times..), in a row, so maybe cron-related ? :
```
 10:45:08
 10:45:08
 10:45:08
 10:45:08
 10:45:08
 11:16:33
 11:16:33
 11:16:33
 12:00:01
 12:00:01
 12:00:01
 12:00:01
 12:00:01
 13:00:27
 13:00:27
 13:00:27
 13:00:27
 13:00:27
 13:36:33
 13:36:33
 13:36:33
 14:15:08
 14:15:08
 14:15:08
 14:15:08
 14:15:08
 15:30:08
 15:30:08
 15:30:08
 15:30:08
```

**Comment by landryb** [comment]
I concur what @reinob said in a previous comment, `occ groufolders:scan 1` is another sure way to trigger the warning/error in the log. Looking at the code flow, the `$folder` object is a groupfolder without 'permissions' entry, as the permissions are per-group:
```
{"id":1,"mount_point":"dir1","groups":{"admin":31,"group1":25},"quota":-3,"size":0}
```

looking a bit more at the `getAllVersionedFiles` method where the error was triggered in 2.0.2, the `$mount` object is only used to gather a list of files, so it doesnt seem that the missing `permissions` entry on the `$folder` object will matter. 

`getAllVersionedFiles` was fixed by commit 5fa67400 in 2.0.3, but the same construct still exists in https://github.com/nextcloud/groupfolders/blob/master/lib/Command/Scan.php#L59 - if i replace `$folder['permissions']` by `$folder['groups']` then  ` occ groupfolders:scan` doesnt trigger the error anymore. @icewind1991 it should probably be fixed this way for the next release ?

There's also the same construct in https://github.com/nextcloud/groupfolders/blob/master/lib/Mount/MountProvider.php#L68 but i'm not sure how this codepath is triggered.

**Comment by icewind1991** [comment]
Should be fixed with https://github.com/nextcloud/groupfolders/commit/77a9a839e07668353a0536cd385ef27be3ea7bf1
